Spigot-Server-Patches/0424-Rate-limit-packets-incoming-from-players.patch

```diff
@@ -0,0 +1,104 @@
+From 5ee916e3f0513eb0f60718f931414278568cf55a Mon Sep 17 00:00:00 2001
+From: Spottedleaf <Spottedleaf@users.noreply.github.com>
+Date: Mon, 11 Mar 2019 12:18:29 -0700
+Subject: [PATCH] Rate limit packets incoming from players
+
+Packets are ratelimited through an interval of time to make sure latency
+spikes don't kill clients. The default limit is 250 packets/second, which
+is also far greater than the max ~70 packets/second vanilla will do according
+to my testing.
+
+Without these ratelimits in place I can happily send 40,000 move packets per
+second which will take down the server.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index a7673dd49..7c4b45522 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -462,4 +462,11 @@ public class PaperConfig {
+         maxBookPageSize = getInt("settings.book-size.page-max", maxBookPageSize);
+         maxBookTotalSizeMultiplier = getDouble("settings.book-size.total-multiplier", maxBookTotalSizeMultiplier);
+     }
++
++    public static int packetRateLimit = 250; // per second
++    public static String packetRateLimitKickMessage = "Sent too many packets";
++    private static void packetRateLimit() {
++        packetRateLimit = Math.max(-1, getInt("settings.spam-limiter.packets-per-second", packetRateLimit));
++        packetRateLimitKickMessage = getString("messages.kick.packet-spam", packetRateLimitKickMessage);
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
+index 30f646e42..de3fb824a 100644
+--- a/src/main/java/net/minecraft/server/NetworkManager.java
++++ b/src/main/java/net/minecraft/server/NetworkManager.java
+@@ -153,6 +153,11 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+     }
+ 
+     private static <T extends PacketListener> void a(Packet<T> packet, PacketListener packetlistener) {
++        // Paper start - Ratelimit packets
```

Comment:
is this same method also called for sync processing? we only need to check in the async handler.

Reply:
This method is asynchronous. We want to prevent flooding of the main thread.

I'm also going to use this method in the next commit to stop packets altogether once we've made the decision to
kill the connection.